pandas/core/reshape/merge.py

```diff
@@ -2346,3 +2384,68 @@ def renamer(x, suffix):
         )
 
     return llabels, rlabels
+
+
+def _chunks(lst, n):
+    for i in range(0, len(lst), n):
+        yield lst[i : i + n]
+
+
+_DEFAULT_LEFT_CHUNK_SIZE = _DEFAULT_RIGHT_CHUNK_SIZE = 1000
+
+
+# TODO: perform lazy merge as optimized cython, rather than chunked merge
+class _LazyCustomInnerMerge:
+    def __init__(
+        self,
+        left: DataFrame | Series,
+        right: DataFrame | Series,
+        # Note: Condition function should respect future desired interface.
+        #       Currently the function "could" make use of df specific logic
+        #       to return a boolean mask (e.g. left.time.dt.date == right.time.dt.date),
+        #       but eventually when this implementation is replaced we might not have
+        #       that ability. Should we restrict those cases now? allow them and fail in
+        #       the future? or plan to develop future implementation to work with these
+        #       constructs as well?
```

Comment:
@jreback - thoughts on this note? My thoughts were that currently since the callable will just be used to boolean mask a chunked dataframe result, with this initial implementation the callable will be able to make use  of dataframe specific constructs, like the `.dt.` and `.str.` accessors. However, when this is eventually cythonized, I'm guessing these features won't be available, not to mention that the comparison targets will no longer be represented by series objects, but rather by scalar values. 

I'm thinking that maybe instead of passing actual df objects to the callable (and allowing access to the underlying series) we instead pass a proxy object that represents the schema of the frames. For now this proxy object would pass through to the actual underlying left/right frames of the chunked result, but with restrictions that would stay consistent through future implementations.

Reply:
@jreback - any thoughts on the above?